Cannot sapply lambda functions

With the cpp11 plugin on, Rcpp cannot sapply lambda functions. 

```cpp
#include <Rcpp.h>
using namespace Rcpp ;

// [[Rcpp:::plugins(cpp11)]]

// [[Rcpp::export]]
NumericVector double_up( NumericVector x ){
  return sapply( x, [](double x){ return x*x; } )
}

/*** R
  double_up( c(1,2) )
**/
```

I get : 

```
$ RcppScript /tmp/double.cpp
In file included from double.cpp:1:
In file included from /Library/Frameworks/R.framework/Versions/3.1/Resources/library/Rcpp/include/Rcpp.h:27:
In file included from /Library/Frameworks/R.framework/Versions/3.1/Resources/library/Rcpp/include/RcppCommon.h:159:
In file included from /Library/Frameworks/R.framework/Versions/3.1/Resources/library/Rcpp/include/Rcpp/traits/traits.h:74:
/Library/Frameworks/R.framework/Versions/3.1/Resources/library/Rcpp/include/Rcpp/traits/result_of.h:31:22: error: no type named 'result_type' in
      '<lambda at double.cpp:8:21>'
        typedef typename T::result_type type ;
                ~~~~~~~~~~~~^~~~~~~~~~~
/Library/Frameworks/R.framework/Versions/3.1/Resources/library/Rcpp/include/Rcpp/sugar/functions/sapply.h:105:34: note: in instantiation of template class
      'Rcpp::traits::result_of<<lambda at double.cpp:8:21> >' requested here
        typename ::Rcpp::traits::result_of<Function>::type ,
                                 ^
double.cpp:8:10: note: while substituting deduced template arguments into function template 'sapply' [with RTYPE = 14, NA = true, T = Rcpp::Vector<14, PreserveStorage>,
      Function = <lambda at double.cpp:8:21>]
  return sapply( x, [](double x){ return x*x; } )
         ^
double.cpp:8:10: error: no matching function for call to 'sapply'
  return sapply( x, [](double x){ return x*x; } )
         ^~~~~~
/Library/Frameworks/R.framework/Versions/3.1/Resources/library/Rcpp/include/Rcpp/vector/ListOf.h:134:3: note: candidate template ignored: could not match 'ListOf' against
      'Vector'
T sapply(const ListOf<T>& t, Function fun) {
  ^
/Library/Frameworks/R.framework/Versions/3.1/Resources/library/Rcpp/include/Rcpp/sugar/functions/sapply.h:109:1: note: candidate template ignored: substitution failure
      [with RTYPE = 14, NA = true, T = Rcpp::Vector<14, PreserveStorage>, Function = <lambda at double.cpp:8:21>]
sapply( const Rcpp::VectorBase<RTYPE,NA,T>& t, Function fun ){
^
2 errors generated.
make: *** [double.o] Error 1
clang++ -std=c++11 -I/Library/Frameworks/R.framework/Resources/include -DNDEBUG  -I/usr/local/include -I/usr/local/include/freetype2 -I/opt/X11/include  -I"/Library/Frameworks/R.framework/Versions/3.1/Resources/library/Rcpp/include"    -fPIC  -O3 -c double.cpp -o double.o
Erreur dans sourceCpp(tail(args, 1), verbose = "-v" %in% args) :
  Error 1 occurred building shared library.
Exécution arrêtée
```

This is mainly `Rcpp::traits:: result_of`'s fault.

Just a heads up, in case this may be of some use: It appears that patching `result_of.h` is not enough.

For testing purposes, I've experimented with replacing the core by the following:
```cpp
#include <type_traits>
namespace Rcpp { namespace traits { using std::result_of; } }
```

This appears to work for `sapply` and a (regular) standalone function, say, `sapply(x, g)` given `double g(double x) { return x; }`.

However, there are still compilation errors for the Romain's lambda example:
```
result_of.cpp: In function 'Rcpp::NumericVector double_up(Rcpp::NumericVector)':
result_of.cpp:23:46: error: no matching function for call to 'sapply(Rcpp::NumericVector&, double_up(Rcpp::NumericVector)::<lambda(double)>)'
result_of.cpp:23:46: note: candidates are:
[...]/R/R-current/library/Rcpp/include/Rcpp/vector/ListOf.h:134:3: note: template<class T, class Function> T Rcpp::sapply(const Rcpp::ListOf<T>&, Function)
[...]/R/R-current/library/Rcpp/include/Rcpp/sugar/functions/sapply.h:109:1: note: template<int RTYPE, bool NA, class T, class Function> Rcpp::sugar::Sapply<RTYPE, NA, T, Function, Rcpp::traits::same_type<typename std::result_of<Function>::type, typename Rcpp::traits::storage_type<Rcpp::traits::r_sexptype_traits<typename std::result_of<Function>::type>::rtype>::type>::value> Rcpp::sapply(const Rcpp::VectorBase<RTYPE, LHS_NA, LHS_T>&, Function)
```

Not quite sure what's the cause here -- could it be that `Rcpp::traits::r_sexptype_traits` or `Rcpp::traits::storage_type` need some further adaptation?

In any case, if making `sapply` work with `std::result_of` is possible, then conditional compilation / dispatching (depending on C++11 support) groundwork has already been done by `boost::result_of`:

* http://www.boost.org/doc/libs/master/boost/utility/result_of.hpp
* http://www.boost.org/doc/libs/master/libs/utility/utility.htm#result_of

On a side note: It's header-only and already a part of the `BH` package. *If* it's relatively self-contained, then perhaps adding a dependency for this particular case wouldn't be too troubling (compared to the current status quo, i.e., not supporting lambdas at all)? In any case, there are only trade-offs...

--- a/inst/include/Rcpp/platform/compiler.h
+++ b/inst/include/Rcpp/platform/compiler.h
@@ -58,6 +58,11 @@
     #endif
 #endif
 
+// Check for the presence of C++0x (or later) support
+#if defined(__GXX_EXPERIMENTAL_CXX0X__) || (__cplusplus >= 201103L)
+  #define RCPP_USING_CXX0X_OR_LATER
+#endif
+
 // Check C++0x/11 features
 #if defined(__INTEL_COMPILER)
     #if __cplusplus >= 201103L

--- a/inst/include/Rcpp/sugar/functions/sapply.h
+++ b/inst/include/Rcpp/sugar/functions/sapply.h
@@ -22,19 +22,42 @@
 #ifndef Rcpp__sugar__sapply_h
 #define Rcpp__sugar__sapply_h
 
+#if (__cplusplus >= 201103L) || defined(__GXX_EXPERIMENTAL_CXX0X__)
+	#define RCPP_USE_CXX11 1
+#else
+	#define RCPP_USE_CXX11 0
+#endif 
+
+#if RCPP_USE_CXX11
+	#include <type_traits>
+#endif
+
 namespace Rcpp{
 namespace sugar{
+
+template <typename Function, typename SugarExpression>
+struct sapply_application_result_of
+{
+#if RCPP_USE_CXX11
+	typedef typename ::std::result_of<Function(typename SugarExpression::stored_type)>::type type;
+#else
+	typedef typename ::Rcpp::traits::result_of<Function>::type type;
+#endif
+} ;
+
+// template <typename Function, typename SugarExpression>
+// using sapply_application_result_of_t = typename sapply_application_result_of<Function, SugarExpression>::type;
 
 template <int RTYPE, bool NA, typename T, typename Function, bool NO_CONVERSION>
 class Sapply : public VectorBase<
 	Rcpp::traits::r_sexptype_traits<
-		typename ::Rcpp::traits::result_of<Function>::type
+		typename ::Rcpp::sugar::sapply_application_result_of<Function, T>::type
 	>::rtype ,
 	true ,
 	Sapply<RTYPE,NA,T,Function,NO_CONVERSION>
 > {
 public:
-	typedef typename ::Rcpp::traits::result_of<Function>::type result_type ;
+	typedef typename ::Rcpp::sugar::sapply_application_result_of<Function, T>::type result_type ;
 	const static int RESULT_R_TYPE =
 		Rcpp::traits::r_sexptype_traits<result_type>::rtype ;
 
@@ -65,13 +88,13 @@ private:
 template <int RTYPE, bool NA, typename T, typename Function>
 class Sapply<RTYPE,NA,T,Function,true> : public VectorBase<
 	Rcpp::traits::r_sexptype_traits<
-		typename ::Rcpp::traits::result_of<Function>::type
+		typename ::Rcpp::sugar::sapply_application_result_of<Function, T>::type
 	>::rtype ,
 	true ,
 	Sapply<RTYPE,NA,T,Function,true>
 > {
 public:
-	typedef typename ::Rcpp::traits::result_of<Function>::type result_type ;
+	typedef typename ::Rcpp::sugar::sapply_application_result_of<Function, T>::type result_type ;
 	const static int RESULT_R_TYPE =
 		Rcpp::traits::r_sexptype_traits<result_type>::rtype ;
 
@@ -102,15 +125,15 @@ template <int RTYPE, bool NA, typename T, typename Function >
 inline sugar::Sapply<
     RTYPE,NA,T,Function,
     traits::same_type<
-        typename ::Rcpp::traits::result_of<Function>::type ,
-        typename Rcpp::traits::storage_type< traits::r_sexptype_traits< typename ::Rcpp::traits::result_of<Function>::type >::rtype >::type
+        typename ::Rcpp::sugar::sapply_application_result_of<Function, T>::type ,
+        typename Rcpp::traits::storage_type< traits::r_sexptype_traits< typename ::Rcpp::sugar::sapply_application_result_of<Function, T>::type >::rtype >::type
     >::value
 >
 sapply( const Rcpp::VectorBase<RTYPE,NA,T>& t, Function fun ){
 	return sugar::Sapply<RTYPE,NA,T,Function,
 	traits::same_type<
-        typename ::Rcpp::traits::result_of<Function>::type ,
-        typename Rcpp::traits::storage_type< traits::r_sexptype_traits< typename ::Rcpp::traits::result_of<Function>::type >::rtype >::type
+        typename ::Rcpp::sugar::sapply_application_result_of<Function, T>::type ,
+        typename Rcpp::traits::storage_type< traits::r_sexptype_traits< typename ::Rcpp::sugar::sapply_application_result_of<Function, T>::type >::rtype >::type
     >::value >( t, fun ) ;
 }
 

--- a/inst/include/Rcpp/sugar/functions/sapply.h
+++ b/inst/include/Rcpp/sugar/functions/sapply.h
@@ -22,14 +22,8 @@
 #ifndef Rcpp__sugar__sapply_h
 #define Rcpp__sugar__sapply_h
 
-#if (__cplusplus >= 201103L) || defined(__GXX_EXPERIMENTAL_CXX0X__)
-	#define RCPP_USE_CXX11 1
-#else
-	#define RCPP_USE_CXX11 0
-#endif 
-
-#if RCPP_USE_CXX11
-	#include <type_traits>
+#if defined(RCPP_USING_CXX0X_OR_LATER)
+	#include <type_traits> // ::std::result_of
 #endif
 
 namespace Rcpp{
@@ -38,7 +32,7 @@ namespace sugar{
 template <typename Function, typename SugarExpression>
 struct sapply_application_result_of
 {
-#if RCPP_USE_CXX11
+#if defined(RCPP_USING_CXX0X_OR_LATER)
 	typedef typename ::std::result_of<Function(typename SugarExpression::stored_type)>::type type;
 #else
 	typedef typename ::Rcpp::traits::result_of<Function>::type type;

